### DISCUSSION TITLE
Is the childrenToReact export from react-markdown stable?

### DISCUSSION THREAD

**Discussion by RyanClementsHax** [remarkjs/remark#1088] [Q&A] [answered]
Hello!

Thanks for this super handy ecosystem. I'm new to the unified space looking to parse markdown and format it on a webpage as html using nextjs. I'm looking to improve performance by doing as much of the parsing/transformation work upfront so I don't have to do it on initial page load in addition to not having to download the dependencies to perform such a task.

This goal led me to want to transform the markdown file through the unified pipeline all the way until it becomes transformed hast and send only that down to the client. The challenge becomes transforming this hast into react components. I see [react-markdown](https://github.com/remarkjs/react-markdown) uses `childrenToReact` [here](https://github.com/remarkjs/react-markdown/blob/4f98f73cca811dc9fb98ee0adce44da981a57b9f/lib/react-markdown.js#L116) in order to do this transformation, but this isn't officially part of the api of the package.

The solution in my head would look like the following:
```js
// on the server

const processor = unified()
    .use(remarkParse)
    .use(remarkFrontmatter)
    .use(remarkRehype)

const hast = await processor.run(processor.parse(rawStringContent))

// on the client

const MyMarkdownComponent = ({ hast }) => {
    const ContentComponent = useMemo(() => childrenToReact(hast), [hast])
    return <ContentComponent />
}
```

For reference, [bayanbennett.com](https://www.bayanbennett.com/) uses this approach for his website as shown [here](https://github.com/BayanBennett/reference.bayanbennett.com/blob/7304c06ff61355501f47b9669a2900a3e601c1f8/src/pages/%5B...path%5D.tsx#L54) and [here](https://github.com/BayanBennett/reference.bayanbennett.com/blob/7304c06ff61355501f47b9669a2900a3e601c1f8/src/pages/%5B...path%5D.tsx#L73)

Consequently I have a few questions:
1. Is this idea dumb/am I missing something?
2. Is there already a tool that does this? The closest thing I can find is [hast-to-hyperscript](https://github.com/syntax-tree/hast-to-hyperscript) but at a glance it seems to do things differently than what y'all do in this package which has me wondering if this is the right solution on top of it not supporting custom component substitutions (from what I can tell)
3. Should I try using [rehype-react](https://github.com/rehypejs/rehype-react) with a noop parser (just one that took in the data and immediately handed it to the transformers) on the client? I tried this approach, but me being new at this was having a difficult time creating such a noop parser (guidance here would be super appreciated). This solution would look like the following:

```js
const MyMarkdownComponent = ({ hast }) => {
    const ContentComponent = useMemo(() => hastToReact(hast), [hast])
    return <ContentComponent />
}

const hastToReact = (hast) => unified()
    .use(function noopParser() {
        return (tree) => tree // I wasn't able to get this working
    })
    .use(rehypeReact, { createElement, Fragment })
    .processSync(hast)
```

Thanks for your time! :)

**Comment by wooorm** [comment, marked as answer]
- It looks like you are implementing React server components yourself, perhaps you can instead use those, which somewhat answers your first question
- unified allows you to a) parse, b) run plugins, and then stop and send things down a write, and the later do c) compile to react: https://github.com/unifiedjs/unified#overview, so this answers your 2nd and 3rd questions: no noops needed, attach the plugins to parse and transform on the server, and plugins to compile on the client

**Reply by RyanClementsHax** [reply]
Thanks for the reply!

Regarding server components, while server components would help, I'd like this to be an SSG solution and correct me if im wrong but server components require SSR. Additionally I am not completely sold on server components for a few reasons and lastly I'm considering possibly using a different framework and want to keep this logic as portable as possible. These were details I left out of the original question so that's my bad.

Regarding the noop parser. I don’t know if I'm doing something wrong but when I only put in the transformer, I get an error saying that a parser is required.

When I try to do something like

```ts
useMemo(
    () =>
      unified()
        .use(rehypeReact, { createElement, Fragment })
        .processSync(content).result,
    [content]
  )
```

I get the error
```
TypeError: Cannot `processSync` without `Parser`
```

And when I try to do something like

```ts
  useMemo(
    () =>
      unified().use(rehypeReact, { createElement, Fragment }).runSync(content),
    [content]
  )
```

It outputs an AST and doesn't do the compile step to create the react nodes. Am I missing a `.compile` function instead of `runSync`? I don't see one on the type definitions

Am I understanding you correctly?

**Reply by wooorm** [reply]
> correct me if im wrong but server components require SSR

This is incorrect. The MDX website is static, with server components.
Though, RSC is still hard to use outside of Next.js, and outside of what the React team thinks is best, they think SSR is best I guess.

> if I'm doing something wrong but when I only put in the transformer, I get an error saying that a parser is required.

That is because you are calling `.process`, which does all three parts: parse, run, stringify. If you want to parse, use `.parse`, if you want to run, use `.run`, if you want to stringify/“compile”, use `.stringify`.

You want to `.parse` and `.run` on the server, send the results down the wire, and `.stringify` on the client.

**Reply by RyanClementsHax** [reply]
Well you learn something new every day. Thanks for the help. .stringify worked.

Marking this as the accepted answer
